src/server/gc_worker/gc_worker.rs

```diff
@@ -235,23 +216,52 @@ impl<E: Engine> GcRunner<E> {
                 break;
             }
 
-            let mut txn = MvccTxn::for_scan(
-                self.engine.snapshot_on_kv_engine(start_key, end_key)?,
-                Some(ScanMode::Forward),
-                TimeStamp::zero(),
-                false,
-            );
-            for key in keys {
-                if let Err(e) = self.gc_key(safe_point, &key, &mut txn) {
-                    error!("gc fail"; "key" => %key, "err" => ?e);
+            fn new_txn<S: Snapshot>(snap: S) -> MvccTxn<S> {
+                MvccTxn::for_scan(snap, Some(ScanMode::Forward), TimeStamp::zero(), false)
+            }
+
+            fn flush_txn<E: Engine>(
+                txn: MvccTxn<E::Snap>,
+                limiter: &Limiter,
+                engine: &E,
+            ) -> Result<()> {
+                let write_size = txn.write_size();
+                let modifies = txn.into_modifies();
+                if !modifies.is_empty() {
+                    limiter.blocking_consume(write_size);
+                    engine.modify_on_kv_engine(modifies)?;
                 }
+                Ok(())
             }
-            let write_size = txn.write_size();
-            let modifies = txn.into_modifies();
-            if !modifies.is_empty() {
-                self.limiter.blocking_consume(write_size);
-                self.engine.modify_on_kv_engine(modifies)?;
+
+            let mut keys = keys.into_iter();
+            let mut txn = new_txn(self.engine.snapshot_on_kv_engine(start_key, end_key)?);
+            let (mut next_key, mut gc_info) = (keys.next(), GcInfo::default());
+            while let Some(ref key) = next_key {
+                self.gc_key(safe_point, key, &mut gc_info, &mut txn);
+                if gc_info.is_completed {
+                    if gc_info.found_versions >= GC_LOG_FOUND_VERSION_THRESHOLD {
+                        debug!(
+                        "GC found plenty versions for a key";
+                        "key" => %key,
+                        "versions" => gc_info.found_versions,
+                        );
+                    }
+                    if gc_info.deleted_versions as usize >= GC_LOG_DELETED_VERSION_THRESHOLD {
+                        debug!(
+                        "GC deleted plenty versions for a key";
+                        "key" => %key,
+                        "versions" => gc_info.deleted_versions,
+                        );
+                    }
+                    next_key = keys.next();
+                    gc_info = GcInfo::default();
```

Comment:
Why not just return the `gc_info` from `gc_key`?

Reply:
If it returns `is_completed = false`, we still need a `GcInfo` outside because we need to merge all `GcInfo`s for one key to print some logs.